src/builtins/string.cpp

```diff
@@ -1,1946 +0,0 @@
-                         const wgetopter_t &w, options_t *opts) {
-    if (opts->all_valid) {
-        opts->all = true;
-        return STATUS_CMD_OK;
-    } else if (opts->allow_empty_valid) {
-        opts->allow_empty = true;
-        return STATUS_CMD_OK;
-    }
-    string_unknown_option(parser, streams, argv[0], argv[w.woptind - 1]);
-    return STATUS_INVALID_ARGS;
-}
-
-static int handle_flag_c(const wchar_t **argv, parser_t &parser, io_streams_t &streams,
-                         const wgetopter_t &w, options_t *opts) {
-    if (opts->chars_to_trim_valid || opts->chars_to_shorten_valid) {
-        opts->chars_to_trim = w.woptarg;
-        return STATUS_CMD_OK;
-    } else if (opts->char_to_pad_valid) {
-        if (wcslen(w.woptarg) != 1) {
-            string_error(streams, _(L"%ls: Padding should be a character '%ls'\n"), argv[0],
-                         w.woptarg);
-            return STATUS_INVALID_ARGS;
-        }
-        opts->char_to_pad = w.woptarg[0];
-        return STATUS_CMD_OK;
-    }
-    string_unknown_option(parser, streams, argv[0], argv[w.woptind - 1]);
-    return STATUS_INVALID_ARGS;
-}
-
-static int handle_flag_e(const wchar_t **argv, parser_t &parser, io_streams_t &streams,
-                         const wgetopter_t &w, options_t *opts) {
-    if (opts->end_valid) {
-        opts->end = fish_wcstol(w.woptarg);
-        if (opts->end == 0 || opts->end == LONG_MIN || errno == ERANGE) {
-            string_error(streams, _(L"%ls: Invalid end value '%ls'\n"), argv[0], w.woptarg);
-            return STATUS_INVALID_ARGS;
-        } else if (errno) {
-            string_error(streams, BUILTIN_ERR_NOT_NUMBER, argv[0], w.woptarg);
-            return STATUS_INVALID_ARGS;
-        }
-        return STATUS_CMD_OK;
-    } else if (opts->entire_valid) {
-        opts->entire = true;
-        return STATUS_CMD_OK;
-    }
-    string_unknown_option(parser, streams, argv[0], argv[w.woptind - 1]);
-    return STATUS_INVALID_ARGS;
-}
-
-static int handle_flag_f(const wchar_t **argv, parser_t &parser, io_streams_t &streams,
-                         const wgetopter_t &w, options_t *opts) {
-    if (opts->filter_valid) {
-        opts->filter = true;
-        return STATUS_CMD_OK;
-    } else if (opts->fields_valid) {
-        for (const wcstring &s : split_string(w.woptarg, L',')) {
-            std::vector<wcstring> range = split_string(s, L'-');
-            if (range.size() == 2) {
-                int begin = fish_wcstoi(range.at(0).c_str());
-                if (begin <= 0 || errno == ERANGE) {
-                    string_error(streams, _(L"%ls: Invalid range value for field '%ls'\n"), argv[0],
-                                 w.woptarg);
-                    return STATUS_INVALID_ARGS;
-                } else if (errno) {
-                    string_error(streams, BUILTIN_ERR_NOT_NUMBER, argv[0], w.woptarg);
-                    return STATUS_INVALID_ARGS;
-                }
```

Comment:
I'm going through this to ensure all error-messages are equivalent. This parsing relies upon the value of the parsed int in the case it fails. That is probably UB? We cannot replicate that from Rust, so this will be a subtle change

Example with current 3.6.1:

```shell
> string split --fields=a "" abc
string split: Invalid fields value 'a'
> string split --fields=1a "" abc
string split: 1a: invalid integer
```

Reply:
Changing the error like that is *fine* - it's awkward how this currently works anyway and error messages are very much not guaranteed to stay the same.

Tho "UB" here would sound like it's undefined behavior in the C++-standard which it isn't - this uses wcstol which returns 0 if no conversion happened or LONG_MIN/LONG_MAX if it was out of range.

Reply:
> Tho "UB" here would sound like it's undefined behavior in the C++-standard which it isn't - this uses wcstol which returns 0 if no conversion happened or LONG_MIN/LONG_MAX if it was out of range.

But the example I gave has `field`/`begin` > 0 when the parsing failed?

Reply:
That's the third case - some conversion happened, but didn't consume all of the string. std::wcstol is happy with this, we set errno to indicate it. In some cases it might be what you want - `math 2pi + 7` could use it (but currently doesn't).

I'm not saying the interface is perfect, mainly my gripe is with the term "UB", which in C++-land is a very specific thing that, as far as I can tell, we do not invoke here.